### ISSUE TITLE
[BUG] [MD] resyncAppState calling before receive appStateSyncKeyShare in first connection

### ISSUE THREAD

**Issue by pokearaujo** [adiwajshing/Baileys#834]
**Describe the bug**
resyncAppState calling before receive appStateSyncKeyShare in first connection, its seen a racing, sometimes came keys before, sometimes not.

**To Reproduce**
1. Create a new connection
2. Probaly will throw "Error: failed to find key \"XXXXXX\" to decode mutation"
3. If not, clear auth, logout and try again

**Expected behavior**
Maybe need wait the keys before request sync collections? Or save in a queue and process after receive the keys

**Comment by adiwajshing** [comment]
Not an issue even if it errors out, doesn't affect the connection

**Comment by ilyaborodin** [comment]
@adiwajshing I think it can affect. Look.

After scanning QR code on my phone, the loader hangs for a long time (about 40-60 seconds) and after that I get an error on my phone that it is not possible to scan qr. Screenshots are attached.

![IMAGE 2021-11-12 13:42:13](https://user-images.githubusercontent.com/48887908/141454170-8f0b89e8-5d30-4cab-b15e-9509f84729a3.jpg)

![IMAGE 2021-11-12 13:42:01](https://user-images.githubusercontent.com/48887908/141454134-dbd91c13-4eaf-4dee-a239-4eb8c90b132d.jpg)


In the terminal, I see that the status "open" comes to me, but the connection is not displayed on my phone. After restarting the connection, the status "open" is also received.

![IMAGE 2021-11-12 13:46:20](https://user-images.githubusercontent.com/48887908/141454687-e341cdde-4922-4613-a3a2-85b80a81dcc0.jpg)

The only problem I see in the terminal is this error, presented below. 
`
{"level":30,"time":1636712914712,"pid":6280,"hostname":"MacBook-Air-Ilya.local","trace":"Error: failed to find key \"AAAAAKmL\" to decode mutation\n    at decodePatches (/Users/ilyaborodin/WebstormProjects/whatsapp_api_md/node_modules/@adiwajshing/baileys-md/lib/Utils/chat-utils.js:275:23)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async resyncAppState (/Users/ilyaborodin/WebstormProjects/whatsapp_api_md/node_modules/@adiwajshing/baileys-md/lib/Socket/chats.js:180:59)","msg":"failed to sync app state"}
`

It happens by accident. But most of the times, I cannot scan the QR code.

**Comment by adiwajshing** [comment]
This isn't related to the app state sync, it's purely client side and doesn't have to do with the server.

Your issue is with the QR, try putting a WA logo in the middle

**Comment by HUGEIT** [comment]
hey man I saw a similar issue happening sometimes with older commits when the initial sync broke before all data was received but at the moment not so much. I do not think it's the logo, I think it's internet connection or something else.

how many chats are on your whatsapp? Are you using the latest commit?

**Comment by HUGEIT** [comment]
@pokearaujo you are right - do you know where exactly the keys are supposed to come in?

is it one of the `setAppStateSyncVersion` references? I think we can use an event emitter to simply notify that we now have the keys needed to request the data. the way it is implemented right now is broken - it will not lead to a disconnect but it is not "complete" either and also the server may expect a request

**Comment by ilyaborodin** [comment]
I don't think this is related to qr. Some times it works, some times no. Also with one phone I can scan successfully but with other no. At this moment I don't have idea. I just yesterday made few tests. And I see if I try to scan via one vpn, I get this error, but if turn off vpn it works. So maybe it is related to internet connection

**Comment by HUGEIT** [comment]
@adiwajshing the issue is still present, your mutex does not work unfortunately. I receive `failed to sync app state` and then only after connecting the second time it can successfully resync and get contacts.

is it possible that when you `throw` it will forget to retry and never do it again or never reach the `critical_unblock_low` in that case?

what I have done now is to simply run this every 10 seconds until it can successfully get the data; not elegant but it works usually the first time the 10 seconds pass.

**Comment by HUGEIT** [comment]
I have made some experimental commit in https://github.com/HUGEIT/Baileys/commit/026ec445a01096bee71169ccf71072623c3f529c basically blocking the mutex until the first set comes in. this generally works in the specified case but when the keys do come in in the now _other_ way around it never finishes the sync and does what @ilyaborodin describes there. when that happens you have a ghost session that is not displayed in the linked devices but still displays all the encrypted messages in the terminal.

basically we need a mix of these two things so this issue stops on both sides. right now it looks like an either-or, either successful login with failed sync very often if the keys needed for critical_block/critical_unblock_low come in late (current status) or successful login in some cases and unsuccessful login in other cases depending on which keys come in first.

**Comment by HUGEIT** [comment]
yes it became worse - these "ghost sessions" block the count of 4 maximum devices, meaning that now I cannot link any more device since there is no way to delete them.

this really needs to be addressed somehow :(

**Comment by adiwajshing** [comment]
@HUGEIT the mutex was never meant to solve this issue -- its only meant to handle parallel modification requests

**Comment by adiwajshing** [comment]
Also let me iterate again -- the appStateSyncKey sync does not affect the connection -- if you dive deeper into the code (`src/sockets/chats.ts`) -- it's only doing some cryptographic checks on the mutations and not transmitting anything back to the server

**Comment by adiwajshing** [comment]
Regardless, https://github.com/adiwajshing/Baileys/commit/72889551b9fd6d484c6507553a7cc92c490b5818 re-syncs on every new key. Should solve this issue. 

Feel free to reopen if the sync does not happen after that as well

**Comment by HUGEIT** [comment]
did not work unfortunately. 

new clone, new scan, Baileys throws `failed to sync app state` and refuses to progress, then WhatsApp after one minute puts up the "Invalid QR code message" that I suppose comes up when the server does not get the client response it expects.

it does not crash, still doing the keep alive stuff but the scan never reaches success in that way.

I will anonymize the log and post it in 5 minutes when I'm done.

**Comment by ilyaborodin** [comment]
I just tested too with new commits. But everything the same like @HUGEIT described above.

**Comment by HUGEIT** [comment]
I think I found the issue. 

is it possible that in 

```
            mutationMutex.mutex(
                () => resyncAppState([ 'critical_block', 'critical_unblock_low' ])
            )
            .catch(err => (
                logger.warn({ trace: err.stack }, 'failed to sync app state')
            ))
```

can someone confirm if that throws/rethrows for the whole block? I don't understand this async await each etc. chaining and its implications but it looks like it could be the reason.

if `critical_block` fails, `critical_unblock_low` never gets reached? the function throwing as a whole instead of a per-argument basis? this adds to my "the WhatsApp server expects a response" thesis. I have not tested it often with my repeated-timeout fix but when I do it, I do not run into this issue. so I think it is possible that one fail makes the whole thing abort. after the first `failed to sync app state` everything stops, I think that is the problem.


so my suggestion, do not make the first parameter throw cancel the chance the second parameter has, and maybe additionally have a retry for the throwing parameters.


in any case, here is the log:



```
{"level":30,"time":1637330073171,"pid":94625,"hostname":" ","jid":" :6@s.whatsapp.net","msg":"registered connection, restart server"}

saving auth state

connection update { isNewLogin: true, qr: undefined }

{"level":30,"time":1637330073172,"pid":94625,"hostname":" ","error":{"data":null,"isBoom":true,"isServer":false,"output":{"statusCode":410,"payload":{"statusCode":410,"error":"Gone","message":"Restart Required"},"headers":{}}},"msg":"connection closed"}
connection update {
  connection: 'close',
  lastDisconnect: {
    error: Error: Restart Required
        at WebSocket.<anonymous> ( Socket/socket.ts:481:17) {
      data: null,
      isBoom: true,
      isServer: false,
      output: [Object]
    },
    date: 2021-11-19T13:54:33.173Z
  }
}

connection update {
  connection: 'connecting',
  receivedPendingNotifications: false,
  qr: undefined
}

{"level":30,"time":1637330073442,"pid":94625,"hostname":" ","msg":"connected to WA Web"}

{"level":20,"time":1637330073598,"pid":94625,"hostname":" ","msg":"handshake recv from WA Web"}

{"level":30,"time":1637330073620,"pid":94625,"hostname":" ","msg":"handshake complete"}

{"level":30,"time":1637330073620,"pid":94625,"hostname":" ","msg":"logging in..."}

{"level":10,"time":1637330074165,"pid":94625,"hostname":" ","fromMe":false,"frame":{"tag":"success","attrs":{"t":"1637330074","props":"26","location":"vll","abprops":"0","creation":"1637330072","companion_enc_static":"I4 ="},"content":null},"msg":"communication"}

{"level":30,"time":1637330074204,"pid":94625,"hostname":" ","msg":"uploaded pre-keys"}

saving auth state

{"level":30,"time":1637330074205,"pid":94625,"hostname":" ","msg":"opened connection to WA"}

{"level":20,"time":1637330074206,"pid":94625,"hostname":" ","msg":"resyncing main app state"}

connection update { connection: 'open' }

{"level":30,"time":1637330074206,"pid":94625,"hostname":" ","msg":"resyncing critical_block from v0"}

{"level":30,"time":1637330074206,"pid":94625,"hostname":" ","msg":"resyncing critical_unblock_low from v0"}

{"level":10,"time":1637330074290,"pid":94625,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"edge_routing","attrs":{},"content":[{"tag":"routing_info","attrs":{},"content":{"type":"Buffer","data":[8,11,8,9]}}]}]},"msg":"communication"}

{"level":10,"time":1637330074291,"pid":94625,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"dirty","attrs":{"type":"account_sync","timestamp":"1637330074"},"content":null}]},"msg":"communication"}

{"level":10,"time":1637330074396,"pid":94625,"hostname":" ","msgId":"8.f-1","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-1"},"content":[{"tag":"active","attrs":{},"content":null}]},"msg":"communication"}

{"level":10,"time":1637330074415,"pid":94625,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"offline","attrs":{"count":"0"},"content":null}]},"msg":"communication"}

{"level":10,"time":1637330074453,"pid":94625,"hostname":" ","msgId":"1476423393","fromMe":false,"frame":{"tag":"notification","attrs":{"from":" @s.whatsapp.net","type":"account_sync","id":"1476423393","t":"1637330074"},"content":[{"tag":"devices","attrs":{"dhash":"2:P o"},"content":[{"tag":"device","attrs":{"jid":" @s.whatsapp.net"},"content":null},{"tag":"device","attrs":{"jid":" :4@s.whatsapp.net","key-index":"1"},"content":null},{"tag":"device","attrs":{"jid":" :5@s.whatsapp.net","key-index":"2"},"content":null},{"tag":"device","attrs":{"jid":" :6@s.whatsapp.net"},"content":null},{"tag":"key-index-list","attrs":{"ts":"1637330071"},"content":{"type":"Buffer","data":[10,. . .137]}}]}]},"msg":"communication"}

{"level":20,"time":1637330074454,"pid":94625,"hostname":" ","recv":{"from":" @s.whatsapp.net","type":"account_sync","id":"1476423393","t":"1637330074"},"sent":{"id":"1476423393","to":" @s.whatsapp.net","class":"notification","type":"account_sync"},"msg":"sent \"notification\" ack"}

{"level":10,"time":1637330074455,"pid":94625,"hostname":" ","msgId":"8.f-2","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-2"},"content":[{"tag":"list","attrs":{},"content":null}]},"msg":"communication"}

{"level":10,"time":1637330074521,"pid":94625,"hostname":" ","msgId":"8.f-0","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-0"},"content":null},"msg":"communication"}

{"level":10,"time":1637330074640,"pid":94625,"hostname":" ","msgId":"8.f-3","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-3"},"content":[{"tag":"privacy","attrs":{},"content":[{"tag":"category","attrs":{"name":"groupadd","value":"all"},"content":null},{"tag":"category","attrs":{"name":"last","value":"all"},"content":null},{"tag":"category","attrs":{"name":"status","value":"all"},"content":null},{"tag":"category","attrs":{"name":"profile","value":"all"},"content":null},{"tag":"category","attrs":{"name":"readreceipts","value":"all"},"content":null}]}]},"msg":"communication"}

{"level":10,"time":1637330074654,"pid":94625,"hostname":" ","msgId":"8.f-4","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-4"},"content":[{"tag":"sync","attrs":{},"content":[{"tag":"collection","attrs":{"name":"critical_block"},"content":[{"tag":"patches","attrs":{},"content":[{"tag":"patch","attrs":{},"content":{"type":"Buffer","data":[10,2. . . ,0,0,234,0]}}]}]}]}]},"msg":"communication"}

{"level":40,"time":1637330074662,"pid":94625,"hostname":" ","trace":"Error: failed to find key \"AAAAAOoA\" to decode mutation\n    at Object.decodePatches (baileys-resyc/src/Utils/chat-utils.ts:330:23)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async resyncAppState ( baileys-resyc/src/Socket/chats.ts:214:59)\n    at async resyncMainAppState ( baileys-resyc/src/Socket/chats.ts:309:9)","msg":"failed to sync app state"}

{"level":10,"time":1637330098819,"pid":94625,"hostname":" ","msgId":"8.f-5","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-5","t":"1637330098"},"content":null},"msg":"communication"}

{"level":10,"time":1637330098820,"pid":94625,"hostname":" ","msg":"recv keep alive"}

{"level":10,"time":1637330123794,"pid":94625,"hostname":" ","msgId":"8.f-6","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"8.f-6","t":"1637330123"},"content":null},"msg":"communication"}

{"level":10,"time":1637330123795,"pid":94625,"hostname":" ","msg":"recv keep alive"}
```

**Comment by adiwajshing** [comment]
Did any of you try to remove the `resyncMainAppState` block from there and see what happens?

**Comment by HUGEIT** [comment]
@adiwajshing do you mean removing it from the chats.ts `ev.on('connection.update', ({ connection }) => {` block?

**Comment by HUGEIT** [comment]
so right now I cannot successfully scan at all anymore no matter what I do. it used to be like an error once in 3 times, now it's a 100% failure rate, similar to what @ilyaborodin describes, nothing happens.

**if I do not comment** I get the error message above.

**if I comment out** the call in the block mentioned above we avoid an error but it simply does not progress from there with the last message of

```
{"level":30,"time":1637340907171,"pid":96657,"hostname":" ","msg":"opened connection to WA"}
connection update { connection: 'open' }
```

and nothing happening after that.

**and now if I put a delay** on the call with the current commit (to give it a bit more time) it goes a bit further without any errors, then fails silently. The first 2 lines are like before, then it progresses but also is like stalling then.

```
{"level":30,"time":1637341112827,"pid":96940,"hostname":" ","msg":"opened connection to WA"}
connection update { connection: 'open' }
{"level":10,"time":1637341113006,"pid":96940,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"edge_routing","attrs":{},"content":[{"tag":"routing_info","attrs":{},"content":{"type":"Buffer","data":[8,11,8,9]}}]}]},"msg":"communication"}
{"level":10,"time":1637341113007,"pid":96940,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"dirty","attrs":{"type":"account_sync","timestamp":"1637341113"},"content":null}]},"msg":"communication"}
{"level":10,"time":1637341113008,"pid":96940,"hostname":" ","msgId":"2.5-1","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"2.5-1"},"content":[{"tag":"active","attrs":{},"content":null}]},"msg":"communication"}
{"level":10,"time":1637341113008,"pid":96940,"hostname":" ","msgId":"2.5-2","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"2.5-2"},"content":[{"tag":"list","attrs":{},"content":null}]},"msg":"communication"}
{"level":10,"time":1637341113018,"pid":96940,"hostname":" ","fromMe":false,"frame":{"tag":"ib","attrs":{"from":"@s.whatsapp.net"},"content":[{"tag":"offline","attrs":{"count":"0"},"content":null}]},"msg":"communication"}
{"level":30,"time":1637341113019,"pid":96940,"hostname":" ","msg":"got 0 offline messages/notifications"}
connection update { receivedPendingNotifications: true }
{"level":10,"time":1637341113020,"pid":96940,"hostname": ","msgId":"2538576377","fromMe":false,"frame":{"tag":"notification","attrs":{"from":" @s.whatsapp.net","type":"account_sync","id":"2---7","t":"1637341113"},"content":[{"tag":"devices","attrs":{"dhash":"2:2 D"},"content":[{"tag":"device","attrs":{"jid":" @s.whatsapp.net"},"content":null},{"tag":"device","attrs":{"jid":" :8@s.whatsapp.net"},"content":null},{"tag":"key-index-list","attrs":{"ts":"1637341108"},"content":{"type":"Buffer","data":[10,18 . . . 146,130]}}]}]},"msg":"communication"}
{"level":20,"time":1637341113021,"pid":96940,"hostname": ","recv":{"from":" @s.whatsapp.net","type":"account_sync","id":"2538576377","t":"1637341113"},"sent":{"id":"2538576377","to":" @s.whatsapp.net","class":"notification","type":"account_sync"},"msg":"sent \"notification\" ack"}
{"level":10,"time":1637341113044,"pid":96940,"hostname": ","msgId":"2.5-3","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"2.5-3"},"content":[{"tag":"privacy","attrs":{},"content":[{"tag":"category","attrs":{"name":"groupadd","value":"all"},"content":null},{"tag":"category","attrs":{"name":"last","value":"all"},"content":null},{"tag":"category","attrs":{"name":"status","value":"all"},"content":null},{"tag":"category","attrs":{"name":"profile","value":"all"},"content":null},{"tag":"category","attrs":{"name":"readreceipts","value":"all"},"content":null}]}]},"msg":"communication"}
{"level":10,"time":1637341113158,"pid":96940,"hostname": ","msgId":"2.5-0","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"2.5-0"},"content":null},"msg":"communication"}
{"level":20,"time":1637341118330,"pid":96940,"hostname": ","msg":"resyncing main app state"}
{"level":30,"time":1637341118331,"pid":96940,"hostname": ","msg":"resyncing critical_block from v0"}
{"level":30,"time":1637341118331,"pid":96940,"hostname": ","msg":"resyncing critical_unblock_low from v0"}
{"level":10,"time":1637341123302,"pid":96940,"hostname": ","msgId":"2.5-4","fromMe":false,"frame":{"tag":"iq","attrs":{"from":"@s.whatsapp.net","type":"result","id":"2.5-4"},"content":[{"tag":"sync","attrs":{},"content":[{"tag":"collection","attrs":{"name":"critical_block"},"content":[{"tag":"patches","attrs":{},"content":null}]},{"tag":"collection","attrs":{"name":"critical_unblock_low"},"content":[{"tag":"patches","attrs":{},"content":null}]}]}]},"msg":"communication"}
saving auth state
```

could some more trace prints around the sync help with that?

**Comment by HUGEIT** [comment]
I set up some more logging on another computer and received 10+ successful scans in a row on the other and then also on the main computer. still the error message in the catch sometimes.

testing another 10 scans I noticed that it never reaches the sync catch after scanning with a fresh leave-beta-and-join-beta. the sync catch only gets entered for all the sessions that come after that.

may there be some problem where baileys is "incompatible" to baileys? that would explain why the first scan does not produce that message.

apart from that, still puzzled why sometimes the scanning does not work at all.

**Comment by itacirgabral** [comment]
> After scanning QR code on my phone, the loader hangs for a long time (about 40-60 seconds) and after that I get an error on my phone that it is not possible to scan qr. Screenshots are attached.

Baileys need connect twice:
https://github.com/adiwajshing/Baileys/blob/multi-device/Example/example.ts#L45

The first connection take the qrcode. While the app are at scanning screen the first connection auto close. To leave the scanning screen the second connection take place and finish the signup. Without the second step the scanning screen hangs and fail.  Could be that what's happening to you?

**Comment by HUGEIT** [comment]
unfortunately that is not the problem - the mere reconnect does not make the the scanning screen dismiss. it is the messages that come after that (or maybe some other message before that we missed)
